### PR TITLE
json configuration

### PR DESCRIPTION
**Pull request by realasberg**
Zum Ausprobieren interessiert dich vermutlich 
    print(options["MangaProject"])
am meisten.
 Das gibt die kompletten geparsten Einstellungsmöglichkeiten zurück.

if __name__ == "__main__":
    path = "./pyManga.json"
    options = parse_jsontemplate(path)
    print(options["MangaProject"])

### CODE REVIEW

File: jsonconfig/pyManga.json [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+                }
+            ]
+        },
+        {
+            "name": "tree_growth_and_death",
+            "$base": "ALTERNATIVE",
+            "typedict": {
+                "alternatives": [
+                    {
+                        "$ref": "tgd:SimpleTest"
```
Comment (jbathmann) [inline jsonconfig/pyManga.json:198]: Why is the parser taking SimpleTest as default?

File: jsonconfig/pyManga.json [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+                    "$base": "STRING",
+                    "value": "NONE"
+                }
+            ]
+        },
+        {
+            "name": "tree_output",
+            "$base": "COMPLEX",
+            "value":[
+                {
```
Comment (jbathmann) [inline jsonconfig/pyManga.json:474]: I don't get it completely yet. Could I add other values here by adding sth like

`
.{"name": ....`

or do I have to switch the base to "alternative"?

File: jsonconfig/jsonoptions.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    return optionMap
+
+
+def buildNode(option):
+    ''' Transform option in xml tree.
+        option.name becomes a xml element tag.
+        option.value becomes element entry if the option is of type "integer", "float" or "string".
+        option.value holds suboptions if option is of type "complex"
+        if option is of type "alternative", option.value is the index of the chosen alternative.
```
Comment (jbathmann) [inline jsonconfig/jsonoptions.py:77]: which alternative will be chosen by default?

File: jsonconfig/jsonoptions.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if option["typedict"]["type"] == "complex":
+        node = ET.Element(name)
+        subnodes = list()
+        for subOption in option["value"]:
+            subnodes.append(buildNode(subOption))
+
+        node.extend(subnodes)
+
+    elif option["typedict"]["type"] == "alternative":
+        index = option["value"]
```
Comment (jbathmann) [inline jsonconfig/jsonoptions.py:89]: Can you explain with this example, what option["value"] does?

**Comment by jbathmann** [comment]
Kuhl. Vielen Dank für den PR! 

Ich hab mal yapf drüber laufen lassen und die doku in unseren Stil gepresst. Hab die commits auf deinem repo als pr geschickt.

**Comment by realasberg** [comment]
Alternativen sind unintiutiv. Die Option.value gibt den index der gewählten alternative an. Default ist 0, also die erste.

Am 14. Juli 2020 23:15:43 MESZ schrieb jbathmann <notifications@github.com>:
>@jbathmann commented on this pull request.
>
>Nice!
>
>I do not understand it yet. Maybe it's getting better, if you could
>answer my (first occurring) questions?
>
>> +                    "value": "FixedSalinity"
>+                },
>+                {
>+                    "name": "salinity",
>+                    "$base": "FLOAT",
>+                    "value": 0.05
>+                }
>+            ]
>+        },
>+        {
>+            "name": "tree_growth_and_death",
>+            "$base": "ALTERNATIVE",
>+            "typedict": {
>+                "alternatives": [
>+                    {
>+                        "$ref": "tgd:SimpleTest"
>
>Why is the parser taking SimpleTest as default?
>
>> +        {
>+            "name": "visualization",
>+            "$base": "COMPLEX",
>+            "value":[
>+                {
>+                    "name": "type",
>+                    "$base": "STRING",
>+                    "value": "NONE"
>+                }
>+            ]
>+        },
>+        {
>+            "name": "tree_output",
>+            "$base": "COMPLEX",
>+            "value":[
>+                {
>
>I don't get it completely yet. Could I add other values here by adding
>sth like
>
>`
>.{"name": ....`
>
>or do I have to switch the base to "alternative"?
>
>
>> +        baseOptions = { data["name"]:data for data in d["base"] }
>+        optionMap = { fqn(data):data for data in d["options"]}
>+
>+        for name,option in optionMap.items():
>+            option =
>option_resolveReferences(option,optionMap,baseOptions)
>+
>+
>+    return optionMap
>+
>+
>+def buildNode(option):
>+    ''' Transform option in xml tree.
>+        option.name becomes a xml element tag.
>+        option.value becomes element entry if the option is of type
>"integer", "float" or "string".
>+        option.value holds suboptions if option is of type "complex"
>+        if option is of type "alternative", option.value is the index
>of the chosen alternative.
>
>which alternative will be chosen by default?
>
>> +        option.name becomes a xml element tag.
>+        option.value becomes element entry if the option is of type
>"integer", "float" or "string".
>+        option.value holds suboptions if option is of type "complex"
>+        if option is of type "alternative", option.value is the index
>of the chosen alternative.
>+    '''
>+    name = option["name"]    
>+    if option["typedict"]["type"] == "complex":
>+        node = ET.Element(name)
>+        subnodes = list()
>+        for subOption in option["value"]:
>+            subnodes.append(buildNode(subOption))
>+
>+        node.extend(subnodes)
>+
>+    elif option["typedict"]["type"] == "alternative":
>+        index = option["value"]
>
>Can you explain with this example, what option["value"] does?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/jbathmann/pyMANGA/pull/49#pullrequestreview-448464738

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

File: jsonconfig/pyManga.json [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+                    "typedict":{
+                        "min": 0,
+                        "max": 500
+                    },
+                    "value": 50
+                }
+            ]
+        },
+        {
+            "name": "domain",
```
Comment (jbathmann) [inline jsonconfig/pyManga.json:357]: why is domain defined twice?

**Comment by realasberg** [comment]
Die eine Domäne hat zusätzliche Parameter. Wenn die zusätzlichen Parameter nur an einer Stelle erlaubt sind, fand ich es am intiutivsten sie zt doppelt zu definieren. Kann auch sein, dass überall alle Parameter erlaubt sind dann könnte man mit minOccurs=0 maxOccurs=1 arbeiten. Das wäre ja eine optionale Einstellung.

Am 15. Juli 2020 15:53:34 MESZ schrieb jbathmann <notifications@github.com>:
>@jbathmann commented on this pull request.
>
>
>
>> +                    },
>+                    "value": 50
>+                },
>+                {
>+                    "name": "y_2",
>+                    "$base": "FLOAT",
>+                    "typedict":{
>+                        "min": 0,
>+                        "max": 500
>+                    },
>+                    "value": 50
>+                }
>+            ]
>+        },
>+        {
>+            "name": "domain",
>
>why is domain defined twice?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/jbathmann/pyMANGA/pull/49#pullrequestreview-448975748

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by jbathmann** [comment]
ich glaub ich checks.
einmal hast du den namespace cmn und einmal cmnres. Zweiteres ist für eine in den Resourcenwettkämpfen definierte domain?

Kannst du die Commits, die ich zu deinem Repo geschickt habe in den PR-Branch mergen? :-)

**Pull request closed without merging**